src/ert/_c_wrappers/enkf/lark_parser.py

```diff
@@ -153,333 +185,89 @@ def substitute(
-            job_name = name
-            arguments = args
-            instruction = JobInstruction(kw, job_name, arguments)
-        else:
-            for node in tree.children[1:]:
-                if node.data != "arg":
-                    raise ConfigValidationError(
-                        "Cannot mix argument list with"
-                        f" parenthesis and without in {node}"
-                    )
-
-                val = node.children[0]
-
-                if not isinstance(val, FileContextToken):
-                    raise ConfigValidationError(
-                        f"Could not read argument {val!r}",
-                        config_file=self.config_file,
-                    )
-                arguments.append(substitute(self.defines, val, expand_env=do_env))
-            instruction = Instruction(kw, arguments)
-        if kw in self.schema:
-            item = self.schema[kw]
-            if (
-                item.argc_min != CONFIG_DEFAULT_ARG_MIN
-                and len(instruction.args) < item.argc_min
-            ):
-                raise ConfigValidationError(
-                    f"{instruction.keyword!r} needs at least {item.argc_min} arguments",
-                    config_file=kw.filename,
-                )
-            if (
-                item.argc_max != CONFIG_DEFAULT_ARG_MAX
-                and len(instruction.args) > item.argc_max
-            ):
-                raise ConfigValidationError(
-                    f"{kw!r} takes maximum {item.argc_max} arguments",
-                    config_file=kw.filename,
-                )
 
-        self.keywords.append(instruction)
+def _substitute_args(args, constraints, defines):
+    if constraints.substitute_from < 1:
+        return args
+    return [
+        _substitute(defines, x, constraints.expand_envvar)
+        if i + 1 >= constraints.substitute_from
+        else x
+        for i, x in enumerate(args)
+    ]
 
 
-def handle_includes(
-    tree: Tree, config_file: str, already_included_files: List[str] = None
+def _handle_includes(
+    tree: Tree[Instruction],
+    defines: List[List[str]],
+    config_file: str,
+    already_included_files: List[str] = None,
 ):
     if already_included_files is None:
         already_included_files = [config_file]
 
     config_dir = os.path.dirname(config_file)
     to_include = []
     for i, node in enumerate(tree.children):
-        if not isinstance(node, Tree):
-            raise ConfigValidationError(
-                f"Unexpected top level statement {node!r}", config_file=config_file
-            )
-        if isinstance(node.children[0], Token):
-            continue  # This is either a newline or a comment
-        if node.children[0].data == "include":
-            inc_node = node.children[0]
-
-            if len(inc_node.children) > 1:
+        kw, *args = node
+        if kw == "DEFINE":
+            constraints = define_keyword()
+            args = constraints.join_args(args)
```

Comment:
Similar comment to that on _tree_to_dict and type annotations. Also, the args are transformed by first join_args, then _substitute_args (but not apply_constraints? I guess you know they are already applied when we are at this point maybe?, but just mentioning it).

Reply:
There are no actual constraints on the defines, except the joining